### PR TITLE
PIM-6460: Attributes removed from family are removed from family variant as well

### PR DESCRIPTION
**Pull request by SamirBoulil**
**Description (for Contributor and Core Developer)**

- Prevents a user from deleting attributes Attributes used as axis in family variants of a family.
- When an attribute is removed from a family, it should be removed from all family variant's attributeSet belonging to this family.
- It should also not appear in the product/product model edit forms.
- Replaced a cascade persist (persist family -> persist family variants) by a dedicated subscriber

----------------
**Reviewable commit by commit**

----------------

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Ok
| Added Behats                      | Ok
| Added integration tests           | Ok
| Changelog updated                 | Ok
| Review and 2 GTM                  | Ok
| Micro Demo to the PO (Story only) | Ok
| CI  | Ok 
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Pim/Bundle/CatalogBundle/tests/integration/Family/UpdateFamilyIntegration.php [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+/**
+ * Tries to update family attributes using imports.
+ *
+ * @author    Samir Boulil <samir.boulil@akeneo.com>
+ * @copyright 2017 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class UpdateFamilyIntegration extends TestCase
+{
+    public function testRemovingSimpleAttributeFromFamilyIsAllowed()
```
Comment (SamirBoulil) [inline src/Pim/Bundle/CatalogBundle/tests/integration/Family/UpdateFamilyIntegration.php]: To remove

File: src/Pim/Bundle/EnrichBundle/Normalizer/FamilyNormalizer.php [hunk 2]
```diff
@@ -5,6 +5,7 @@
 use Pim\Bundle\CatalogBundle\Filter\CollectionFilterInterface;
```
Comment (SamirBoulil) [inline src/Pim/Bundle/EnrichBundle/Normalizer/FamilyNormalizer.php]: add declare

File: src/Pim/Bundle/CatalogBundle/EventSubscriber/RemoveAttributeFromFamilyVariantsOnFamilyUpdateSubscriber.php [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+/**
+ * Whenever an attribute is removed from a family, we need to ensure this attribute is removed from every family
+ * variants belonging to this family.
+ *
+ * @author    Samir Boulil <samir.boulil@akeneo.com>
+ * @copyright 2017 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class RemoveAttributeFromFamilyVariantsOnFamilyUpdateSubscriber implements EventSubscriberInterface
```
Comment (ahocquard) [inline src/Pim/Bundle/CatalogBundle/EventSubscriber/RemoveAttributeFromFamilyVariantsOnFamilyUpdateSubscriber.php]: Attributes?

File: src/Pim/Bundle/CatalogBundle/EventSubscriber/RemoveAttributeFromFamilyVariantsOnFamilyUpdateSubscriber.php [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+            StorageEvents::PRE_SAVE => 'removeDeletedAttributeFromFamilyVariants',
+        ];
+    }
+
+    /**
+     * Removes the removed attributes from family from the family variants belonging to this family.
+     *
+     * @param GenericEvent $event
+     */
+    public function removeDeletedAttributeFromFamilyVariants(GenericEvent $event): void
```
Comment (ahocquard) [inline src/Pim/Bundle/CatalogBundle/EventSubscriber/RemoveAttributeFromFamilyVariantsOnFamilyUpdateSubscriber.php]: attributes?

File: src/Pim/Bundle/CatalogBundle/tests/integration/Family/UpdateFamilyIntegration.php [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+    {
+        $violations = $this->removeAttributeFromFamily('shoes', ['material']);
+        $this->assertEquals(0, $violations->count());
```
Comment (ahocquard) [inline src/Pim/Bundle/CatalogBundle/tests/integration/Family/UpdateFamilyIntegration.php]: please assert on property path as well, it is used in the API.

https://github.com/akeneo/pim-community-dev/blob/2.0/src/Pim/Bundle/ApiBundle/tests/integration/Controller/Family/PartialUpdateFamilyIntegration.php#L462

(no need to add a test in the API, just ensure that the correct property path is set).

File: src/Pim/Component/Catalog/Validator/Constraints/FamilyAttributeUsedAsAxisValidator.php [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+     * @param AttributeInterface[]   $missingAttributeCodesUsedAsAxis
+     */
+    private function buildViolationsForMissingAttributesUsedAsAxis(
+        Constraint $constraint,
+        FamilyVariantInterface $familyVariant,
+        array $missingAttributeCodesUsedAsAxis
+    ): void {
+        foreach ($missingAttributeCodesUsedAsAxis as $missingAttributeUsedAsAxis) {
+            $this->context
+                ->buildViolation($constraint->messageAttribute, [
```
Comment (ahocquard) [inline src/Pim/Component/Catalog/Validator/Constraints/FamilyAttributeUsedAsAxisValidator.php]: specify the property path

File: src/Pim/Component/Catalog/Validator/Constraints/FamilyAttributeUsedAsAxisValidator.php [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+ */
+class FamilyAttributeUsedAsAxisValidator extends ConstraintValidator
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function validate($family, Constraint $constraint)
+    {
+        if (!$family instanceof FamilyInterface) {
+            return;
```
[the remark this answers is not in the document]
Comment (SamirBoulil) [reply src/Pim/Component/Catalog/Validator/Constraints/FamilyAttributeUsedAsAxisValidator.php]: nope, they only return

File: src/Pim/Component/Catalog/Validator/Constraints/FamilyAttributeUsedAsAxisValidator.php (hunk 6 above)
Comment (SamirBoulil) [reply src/Pim/Component/Catalog/Validator/Constraints/FamilyAttributeUsedAsAxisValidator.php]: done

File: src/Pim/Bundle/CatalogBundle/tests/integration/Family/UpdateFamilyIntegration.php (hunk 5 above)
Comment (SamirBoulil) [reply src/Pim/Bundle/CatalogBundle/tests/integration/Family/UpdateFamilyIntegration.php]: done

File: src/Pim/Bundle/CatalogBundle/EventSubscriber/RemoveAttributeFromFamilyVariantsOnFamilyUpdateSubscriber.php (hunk 4 above)
Comment (SamirBoulil) [reply src/Pim/Bundle/CatalogBundle/EventSubscriber/RemoveAttributeFromFamilyVariantsOnFamilyUpdateSubscriber.php]: done

File: src/Pim/Bundle/CatalogBundle/EventSubscriber/RemoveAttributeFromFamilyVariantsOnFamilyUpdateSubscriber.php (hunk 3 above)
Comment (SamirBoulil) [reply src/Pim/Bundle/CatalogBundle/EventSubscriber/RemoveAttributeFromFamilyVariantsOnFamilyUpdateSubscriber.php]: done

File: src/Pim/Component/Catalog/Validator/Constraints/FamilyAttributeUsedAsAxisValidator.php [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+ * @author    Samir Boulil <samir.boulil@akeneo.com>
+ * @copyright 2017 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class FamilyAttributeUsedAsAxisValidator extends ConstraintValidator
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function validate($family, Constraint $constraint)
```
[the remark this answers is not in the document]
Comment (SamirBoulil) [reply src/Pim/Component/Catalog/Validator/Constraints/FamilyAttributeUsedAsAxisValidator.php]: done

File: src/Pim/Component/Catalog/Validator/Constraints/FamilyAttributeUsedAsAxisValidator.php [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+namespace Pim\Component\Catalog\Validator\Constraints;
+
+use Pim\Component\Catalog\Model\AttributeInterface;
+use Pim\Component\Catalog\Model\FamilyInterface;
+use Pim\Component\Catalog\Model\FamilyVariantInterface;
+use Symfony\Component\Validator\Constraint;
+use Symfony\Component\Validator\ConstraintValidator;
+
+/**
+ *  Family attribute_used_as_axis constraint
```
[the remark this answers is not in the document]
Comment (SamirBoulil) [reply src/Pim/Component/Catalog/Validator/Constraints/FamilyAttributeUsedAsAxisValidator.php]: done

File: src/Pim/Component/Catalog/Validator/Constraints/FamilyAttributeUsedAsAxis.php [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+ */
+class FamilyAttributeUsedAsAxis extends Constraint
+{
+    /** @var string */
+    public $messageAttribute = 'Attribute "%attribute%" is an axis in "%family_variant%" family variant. It must belong to the family.';
+
+    /**
+     * {@inheritdoc}
+     */
+    public function validatedBy()
```
[the remark this answers is not in the document]
Comment (SamirBoulil) [reply src/Pim/Component/Catalog/Validator/Constraints/FamilyAttributeUsedAsAxis.php]: done

File: src/Pim/Component/Catalog/Validator/Constraints/FamilyAttributeUsedAsAxis.php [hunk 11]
```diff
@@ -0,0 +1,35 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Pim\Component\Catalog\Validator\Constraints;
+
+use Symfony\Component\Validator\Constraint;
+
+/**
+ *
```
[the remark this answers is not in the document]
Comment (SamirBoulil) [reply src/Pim/Component/Catalog/Validator/Constraints/FamilyAttributeUsedAsAxis.php]: done

**Pull request merged**